realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -0,0 +1,388 @@
+    public void close() {
+        checkIfValidThread();
+        // If Realms are still being opened, abort that task
+        if (openInProgress) {
+            if (managementRealmOpenTask != null) {
+                managementRealmOpenTask.cancel();
+                managementRealmOpenTask = null;
+            }
+            if (permissionRealmOpenTask != null) {
+                permissionRealmOpenTask.cancel();
+                permissionRealmOpenTask = null;
+            }
+        } else {
+            if (managementRealm != null) {
+                managementRealm.close();
+            }
+            if (permissionRealm != null) {
+                permissionRealm.close();
+            }
+        }
+        closed = true;
+    }
+
+    /**
+     * Checks if this PermissionManager is closed or not. If it is closed, all methods will report back an error
+     *
+     * @return {@code true} if the PermissionManager is closed, {@code false} if it is still open.
+     */
+    public boolean isClosed() {
+        checkIfValidThread();
+        return closed;
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
+        super.finalize();
+        if (managementRealm != null || permissionRealm != null) {
+            RealmLog.error("PermissionManager was not correctly closed before being finalized.");
+            close();
+        }
+    }
+
+    // Creates the URL to the permission/management Realm based on the authentication URL.
+    private static String getRealmUrl(RealmType type, URL authUrl) {
+        String scheme = "realm";
+        if (authUrl.getProtocol().equalsIgnoreCase("https")) {
+            scheme = "realms";
+        }
+        try {
+            return new URI(scheme, authUrl.getUserInfo(), authUrl.getHost(), authUrl.getPort(),
+                    "/~/" + type.getName(), null, null).toString();
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("Could not create URL to the " + type + " + Realm", e);
+        }
+    }
+
+    // Task responsible for loading the Permissions result and returning it to the user.
+    // The Permission result is not considered available until the query has complated.
+    private class PermissionsAsyncTask extends PermissionManagerAsyncTask {
+
+        private final Callback<RealmResults<Permission>> callback;
+        private RealmResults<Permission> loadingPermissons; // Prevent permissions from being GC'ed until fully loaded.
+
+        PermissionsAsyncTask(Callback<RealmResults<Permission>> callback) {
+            this.callback = callback;
+        }
+
+        @Override
+        public void run() {
+            if (isCancelled()) { return; }
+            if (permissions != null || isClosed()) {
+                notifyCallbackWithResult();
+            } else {
+                synchronized (errorLock) {
+                    if (permissionRealmError != null) {
+                        callback.onError(permissionRealmError);
```

Comment:
This seems dangerous to me.  You are invoking the callback while holding the lock.  Would it be better to get a reference to the lock and then do the callback outside the synchronized block?